src/item.cpp

```diff
@@ -4029,11 +4029,18 @@ bool item::ammo_consume( int qty ) {
 
     item *mag = magazine_current();
     if( mag ) {
-        auto res = mag->ammo_consume( qty );
-        if( res && ammo_remaining() == 0 && mag->has_flag( "NO_RELOAD" ) ) {
-            contents.erase( std::remove_if( contents.begin(), contents.end(), [&mag]( const item& e ) {
-                return mag == &e;
-            } ) );
+        auto res = mag->ammo_consume( qty, pos );
+        if( res && ammo_remaining() == 0 ) {
+            if( mag->has_flag( "MAG_DESTROY" ) ) {
+                contents.erase( std::remove_if( contents.begin(), contents.end(), [&mag]( const item& e ) {
+                    return mag == &e;
+                } ) );
+            } else if ( mag->has_flag( "MAG_EJECT" ) ) {
+                g->m.add_item( pos, *mag );
```

Comment:
Note for later, might be worthwhile to add a map method like drop_near( pos, item, range ) that finds an open nearby square and drops the item there, good for this and shell casings, and probably some other things.